File: retrofit2_rxjava2/src/main/java/jack/retrofit2_rxjava2/manager/rx/RxFunction.java
package jack.retrofit2_rxjava2.manager.rx;

import java.io.IOException;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;
import jack.retrofit2_rxjava2.exception.ApiException;
import jack.retrofit2_rxjava2.exception.DataNullException;
import jack.retrofit2_rxjava2.exception.TimeOutException;
import jack.retrofit2_rxjava2.exception.UnloginException;
import jack.retrofit2_rxjava2.model.ApiResponse;
import jack.retrofit2_rxjava2.util.net.NetConfig;

/**
 * created by Jack
 * email:yucrun@163.com
 * date:19-5-5
 * describe:将ApiResponse<T>转换成T格式
 *
 * 数据剥壳：同时对stat的不同状态进行处理
 *
 * （实践发现的）问题：在飞行模式的情况下，RxFunction的代码不执行。具体原因不清楚(夜神模拟器和手机均测试)
 */
public final class RxFunction<T> implements Function<ApiResponse<T>, T> {

    @Override
    public T apply(@NonNull ApiResponse<T> apiResponse) throws IOException {
            int status = apiResponse.getErrorCode();

            System.out.println(" RxFunction status " + status);

            if(status == NetConfig.CODE_ERROR || status == NetConfig.NOT_MATCH){
                System.out.println(" RxFunction " + apiResponse.getErrorMsg());
                throw new ApiException(apiResponse.getErrorCode(),apiResponse.getErrorMsg());
            }else if(status == NetConfig.UN_LOGIN){
                System.out.println(" RxFunction " + apiResponse.getErrorMsg());
                throw new UnloginException(apiResponse.getErrorCode(),apiResponse.getErrorMsg());
            }else if(status == NetConfig.CODE_SUCCESS){
                if (apiResponse.getData() == null) {
                    System.out.println(" RxFunction " + apiResponse.getErrorMsg());
                    throw new DataNullException(apiResponse.getErrorCode(),apiResponse.getErrorMsg());
                }
                return apiResponse.getData();
            }else {
                System.out.println(" RxFunction " + apiResponse.getErrorMsg());
                throw new TimeOutException(apiResponse.getErrorCode(), "请求超时");
            }

    }

}

File: retrofit2_rxjava2/src/main/java/jack/retrofit2_rxjava2/manager/rx/RxFunction.kt
package jack.retrofit2_rxjava2.manager.rx

import io.reactivex.functions.Function
import jack.retrofit2_rxjava2.util.net.NetConfig
import jack.retrofit2_rxjava2.exception.ApiException
import jack.retrofit2_rxjava2.exception.UnloginException
import jack.retrofit2_rxjava2.exception.DataNullException
import jack.retrofit2_rxjava2.exception.TimeOutException
import jack.retrofit2_rxjava2.model.ApiResponse
import java.io.IOException

/**
 * created by Jack
 * date:19-5-5
 * describe:将ApiResponse<T>转换成T格式
 *
 * 数据剥壳：
 * 同时对stat的不同状态进行处理
 * 执行rxfuntion的前提，是后台能返回apiResponse的数据
</T> */
class RxFunction<T> : Function<ApiResponse<T>, T> {
    @Throws(IOException::class)
    override fun apply(apiResponse: ApiResponse<T>): T {
        val status = apiResponse.errorCode
        return if (status == NetConfig.CODE_ERROR || status == NetConfig.NOT_MATCH) {
            throw ApiException(apiResponse.errorCode, apiResponse.errorMsg)
        } else if (status == NetConfig.UN_LOGIN) {
            throw UnloginException(apiResponse.errorCode, apiResponse.errorMsg)
        } else if (status == NetConfig.CODE_SUCCESS) {
            if (apiResponse.data == null) {
                //这里通过自定义转换器 对 返回的数据为null的情况 做一下处理
                throw DataNullException(apiResponse.errorCode, apiResponse.errorMsg)
            }
            apiResponse.data
        } else {
            throw TimeOutException(apiResponse.errorCode, "请求超时")
        }
    }
}